src/modules/rover_pos_control/RoverPositionControl.cpp

```diff
@@ -380,31 +420,57 @@ RoverPositionControl::control_attitude(const vehicle_attitude_s &att, const vehi
 	// quaternion attitude control law, qe is rotation from q to qd
 	const Quatf qe = Quatf(att.q).inversed() * Quatf(att_sp.q_d);
 	const Eulerf euler_sp = qe;
+	// PX4_INFO("Yaw error: %f", double(euler_sp(2)));
+
+	_rates_sp.roll = 0.0;
+	_rates_sp.pitch = 0.0;
+	_rates_sp.yaw = euler_sp(2) / _param_max_turn_angle.get();
+	_rates_sp.thrust_body[0] = math::constrain(att_sp.thrust_body[0], 0.0f, 1.0f);
+	_rates_sp.timestamp = hrt_absolute_time();
+	_rates_sp_pub.publish(_rates_sp);
+}
 
-	float control_effort = euler_sp(2) / _param_max_turn_angle.get();
-	control_effort = math::constrain(control_effort, -1.0f, 1.0f);
+void
+RoverPositionControl::control_rates(const vehicle_angular_velocity_s &rates, const  vehicle_angular_acceleration_s &acc,
+				    const vehicle_local_position_s &local_pos,
+				    const vehicle_rates_setpoint_s &rates_sp)
+{
+	float dt = (_control_rates_last_called > 0) ? hrt_elapsed_time(&_control_rates_last_called) * 1e-6f : 0.01f;
+	_control_rates_last_called = hrt_absolute_time();
 
-	_act_controls.control[actuator_controls_s::INDEX_YAW] = control_effort;
+	const matrix::Vector3f vehicle_rates(rates.xyz[0], rates.xyz[1], rates.xyz[2]);
+	const matrix::Vector3f rates_setpoint(rates_sp.roll, rates_sp.pitch, rates_sp.yaw);
 
-	const float control_throttle = att_sp.thrust_body[0];
+	const matrix::Vector3f current_velocity(local_pos.vx, local_pos.vy, local_pos.vz);
+	bool lock_integrator = bool(current_velocity.norm() < _param_rate_i_minspeed.get());
 
-	_act_controls.control[actuator_controls_s::INDEX_THROTTLE] =  math::constrain(control_throttle, 0.0f, 1.0f);
+	const matrix::Vector3f angular_acceleration{acc.xyz};
+	const matrix::Vector3f torque = _rate_control.update(vehicle_rates, rates_setpoint, angular_acceleration, dt,
+					lock_integrator);
 
+	_steering_input = math::constrain(_steering_input + torque(2), -1.0f, 1.0f);
```

Comment:
Shouldn't the steering input be set directly by the torque? Why would we accumulate the torque setpoint?

For example when we have achieved our desired rate, the FF gain will then still be accumulating in the steering input, which isn't what we want.

Reply:
Not really, the angular acceleration maps to the steering rate and not the steering input

Reply:
How come it's the steering rate? When you want to apply certain torque, shouldn't the rudder deflection (steering) be proportional to that torque setpoint, instead of being integral of it?

Reply:
If we assume $yaw_rate = velocity * steering$
$yaw_acceleration = acceleration * steering + velocity * steering_rate$

Reply:
> If we assume 
> ψ˙=V∗δ where δ is the steering and 
> ψ˙ is the yaw rate.

What does V stand for?

> If rudder input directly corresponds to torque(or angular acceleration), with a given rudder and speed input you should have the yaw rate accelerating faster and faster. However, it normally converges quickly to a certain yaw rate.

Shouldn't this be handled by the I-term of the PID controller? Or, regarding the damping, is integrating the steering input the common way to mitigate this torque control problem?

Reply:
> What does V stand for?

V stands for velocity

> Shouldn't this be handled by the I-term of the PID controller? Or, regarding the damping, is integrating the steering input the common way to mitigate this torque control problem?

I don't think it should be handled by the PID controller assuming steering corresponds to torque (because it does not)

Reply:
@junwoo091400 If you think using a torque model is the right way, I am all open for it as long as you can show that it is a better way to handle rate control. However, I think this is already an improvement of the previous state and worth merging.

Reply:
For now I have temporarily pushed this single commit to address this change I proposed, but we can revert back, if it proves to be inefficient, after real life testing of the vehicle: https://github.com/PX4/PX4-Autopilot/pull/20082/commits/205920e013f2d5d3a5f7d09d3974c819db82ee3c

Reply:
@junwoo091400 Here is what ardupilot is doing: https://github.com/ArduPilot/ardupilot/blob/f74d22135cc23b36f142c955314408fb3adc566d/libraries/APM_Control/AR_AttitudeControl.cpp#L551

Reply:
I think you are probably right, ardurover is also on a pid loop, but with ff https://github.com/ArduPilot/ardupilot/blob/f74d22135cc23b36f142c955314408fb3adc566d/libraries/APM_Control/AR_AttitudeControl.cpp#L572-L573